### PR TITLE
Add util function to delete supervisor cluster snaphsot cr explicitly

### PR DESCRIPTION
**Pull request by xinyanw409**
Add a utility function so that when deleting snapshot CR in guest cluster, the corresponding supervisor cluster snapshot CR can also be deleted explicitly.

Added unit tests for this function.
```
Running tests: ./pkg/utils/
ok  	github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/utils	0.063s
Success!
```

Precheck: https://container-dp.svc.eng.vmware.com/job/Container_Precheck_Velero/589/

Related bug link: https://bugzilla.eng.vmware.com/show_bug.cgi?id=2657386

### CODE REVIEW

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/294/

**Comment by svcbot-qecnsdp** [comment]
Congratulations !! 

All the pre-check tests passed
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/295/

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -904,3 +908,31 @@ func IsObjectBlocked(item runtime.Unstructured) (bool, string, error) {
...
+	pluginClient, err := plugin_clientset.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return pluginClient.BackupdriverV1alpha1(), nil
+}
+
+// Provide a utility function in guest cluster to clean up corresponding supervisor cluster snapshot CR
+func DeleteSvcSnapshot(gcSnapshot *backupdriverapi.Snapshot, svcConfig *rest.Config, svcNamespace string, logger logrus.FieldLogger) error {
+	svcSnapshotName := gcSnapshot.Status.SvcSnapshotName
```
Comment (xing-yang) [inline pkg/utils/utils.go:922]: Add a check here to see if gcSnapshot.Status.SvcSnapshotName is an empty string.  If it is empty string, return here.

**Comment by xing-yang** [comment]
Have you tested this on a Guest Cluster?

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -904,3 +908,34 @@ func IsObjectBlocked(item runtime.Unstructured) (bool, string, error) {
...
+	if svcSnapshotName == "" {
+		return errors.New("No corresponding supervisor cluster snapshot CR.")
+	}
+	backupdriverClient, err := GetBackupdriverClient(svcConfig)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to retrieve plugin client from svcConfig %v", svcConfig)
+		return err
+	}
+	err = backupdriverClient.Snapshots(svcNamespace).Delete(context.TODO(), svcSnapshotName, metav1.DeleteOptions{})
```
Comment (lintongj) [inline pkg/utils/utils.go:931]: Suggestion here.
if err = ...; k8serrors.IsNotFound(err) {
   ...
} else if err  != nil {
   ...
}

File: pkg/utils/utils_test.go [hunk 3]
```diff
@@ -337,3 +342,94 @@ func Test_waitForPvSecret(t *testing.T) {
...
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "test-ns",
+					Name:      "svc-snapshot-2",
+				},
+			},
+			config: &rest.Config{},
+			toDeleteSvcSnapshotFirst: false,
+			expectedErr: nil,
+		},
```
Comment (lintongj) [inline pkg/utils/utils_test.go:412]: it would be better to add some negative cases.

**Review by lintongj — COMMENTED** [review]
General question:  which component will consume the newly introduced util function? @xinyanw409

**Comment by xinyanw409** [comment]
> General question: which component will consume the newly introduced util function? @xinyanw409

Currently our plugin code will not use this util function. We just provide a method for PowerProtect so they can clean up the supervisor cluster snapshot CR when they explicitly delete guest cluster snapshot CR.
Related bug number: https://bugzilla.eng.vmware.com/show_bug.cgi?id=2657386

**Comment by svcbot-qecnsdp** [comment]
Congratulations !! 

All the pre-check tests passed
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/296/

**Review by lintongj — APPROVED** [review]
/lgtm

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/297/

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -904,3 +908,32 @@ func IsObjectBlocked(item runtime.Unstructured) (bool, string, error) {
...
+		return nil, err
+	}
+	return pluginClient.BackupdriverV1alpha1(), nil
+}
+
+// Provide a utility function in guest cluster to clean up corresponding supervisor cluster snapshot CR
+func DeleteSvcSnapshot(gcSnapshot *backupdriverapi.Snapshot, svcConfig *rest.Config, svcNamespace string, logger logrus.FieldLogger) error {
+	svcSnapshotName := gcSnapshot.Status.SvcSnapshotName
+	if svcSnapshotName == "" {
+		return errors.New("No corresponding supervisor cluster snapshot CR.")
```
Comment (xing-yang) [inline pkg/utils/utils.go:924]: Log an error message.

**Comment by svcbot-qecnsdp** [comment]
Congratulations !! 

All the pre-check tests passed
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/298/

**Comment by swgupta-vmw** [comment]
What is the order of calling this function? My understanding is the guest snapshot CR should be deleted before the supervisor snapshot CR. This is to avoid the the guest controller from recreating the supervisor snapshot CR after it is deleted. But in that case, this utility function will not work as it takes the guest snapshot CR as input.

Should we consider updating the utility function to have the following signature:

func DeleteSvcSnapshot(svcSnapshotName string, guestConfig *rest.Config, logger logrus.FieldLogger) error

The caller can save the svcSnapshotName from the guest snapshot CR before deleting it.
Entering guestConfig is easier for PowerProtect. We can call GetSupervisorConfig() internally to get svc config and namespace.

**Comment by xing-yang** [comment]
Inside DeleteSvcSnapshot, gcSnapshot *backupdriverapi.Snapshot is not retrieved from API server again.   It is just a copy of the Snapshot object.  So this should work even if the Snapshot CR is deleted in GC first.

**Comment by swgupta-vmw** [comment]
ok. I guess it will work. It just seems cleaner if we take the svcSnapshotName string as input since the CR itself has been deleted.

**Comment by xinyanw409** [comment]
> ok. I guess it will work. It just seems cleaner if we take the svcSnapshotName string as input since the CR itself has been deleted.

Based on offline discussion with Swati, I agree that it will be better to take svcSnapshotName and guest config as input here so that it will be easier for PowerProtect to use.  Make a following PR https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/220.

**Pull request merged**